File: algo/78. Subsets.js
// @9 DFS

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    
    var helper = function(res, collection, i) {
        // out of bound
        var collection = Array.from(collection);       
        if (i > nums.length) {
            return;
        }
        res.push(collection);
        
        for (var index = i; index < nums.length; index++) {
            collection.push(nums[index]);
            helper(res, collection, index+1);
            collection.pop();
        }
        
    }
    
    var res = [];
    if (nums.length == 0) {
        return [];
    }
    nums.sort();
    helper(res, [], 0);
        
    return res;
};

File: round2/78. Subsets.cs
public class Solution {
    private void DFS(List<IList<int>> res, List<int> curr, int index, int[] nums) {
        res.Add(new List<int>(curr));
        for (int i=index; i<nums.Length; i++) {
            curr.Add(nums[i]);
            DFS(res, curr, i+1, nums);
            curr.RemoveAt(curr.Count - 1);
        }
    }
    public IList<IList<int>> Subsets(int[] nums) {
        var res = new List<IList<int>>();
        var curr = new List<int>();
        DFS(res, curr, 0, nums);
        return res;
    }
}